js/viewer/annotation/text.js

```diff
@@ -0,0 +1,425 @@
+    pText.addEventListener("mousemove", _self.onMousemoveShowTooltip);
+    pText.addEventListener("mouseout", _self.onMouseoutHideTooltip);
+
+    new OpenSeadragon.MouseTracker({
+        element: pText,
+        clickHandler: function (e) {
+            // intentially left empy. we just want to prevent the default behavior (zoom)
+        }
+    });
+
+    return pText;
+}
+
+/**
+ * This function is used to remove the text annotation from the OSD canvas
+ * and also remove the event listeners attached to the <p> tag.
+ */
+Text.prototype.removeText = function () {
+
+    var foreignObj = this.getForeignObj();
+    if(foreignObj == null) return;
+    var pText = foreignObj.getElementsByTagName("p")[0];
+    if(pText == null) return;
+
+    pText.removeEventListener("click", this.onClickToSelectAnnotation);
+    pText.removeEventListener("mouseover", this.onMouseoverShowTooltip);
+    pText.removeEventListener("mouseout", this.onMouseoutHideTooltip);
+    foreignObj.parentNode.removeChild(foreignObj);  
+}
+
+/**
+ * This function adds a textbox to the SVG group. It adds a foreignObject > div > textarea.
+ * If an importedObj exists, the function replaces the foreign object with the foreign object
+ * of imported annotation.
+ * @param {*} groupId 
+ * @param {*} importedObj - The foreign object from the imported annotation.
+ * @returns 
+ */
+Text.prototype.addTextBox = function (groupId, importedObj) {
+
+    // If a foreign object is to be imported from the loaded annotation
+    if(importedObj != null){
+        this.setForeignObj(importedObj);
+        var pTag = this.createPTag(importedObj.childNodes[0]);
+        importedObj.appendChild(pTag);
+        group = document.getElementById(groupId);
+        group.appendChild(importedObj);
+        group.setAttribute("contentEditable", "true");
+        return;
+    }
+
+    var svgForeignObj = document.createElementNS('http://www.w3.org/2000/svg', 'foreignObject')
+    var textOuterDiv = document.createElementNS("http://www.w3.org/1999/xhtml", "div");
+    var textInput = document.createElementNS("http://www.w3.org/1999/xhtml", "textarea");
+    this.setForeignObj(svgForeignObj);
+    
+    fontSize = document.getElementsByClassName("fontInput")[0].value;
+    this.changeFontSize(fontSize);
+
+    // Set the attributes of the foreign object, div and textarea
+    textInput.setAttribute("id", "textInput");
+    textOuterDiv.setAttribute("contentEditable", "true");
+    textInput.setAttribute("contentEditable", "true");
+    textInput.style.color = this._attrs["stroke"];
+    svgForeignObj.setAttribute("contentEditable", "true");
+    textOuterDiv.setAttribute("width", "auto");
+    textOuterDiv.setAttribute("tabindex", "-1");
+    svgForeignObj.setAttribute("width", "100%");
+    svgForeignObj.setAttribute("height", this._attrs["font-size"] * 100 + "px");
+    textOuterDiv.appendChild(textInput); 
+    svgForeignObj.classList.add("foreign-object");
```

Comment:
Why are you adding these classes?

Reply:
These classes are necessary for styling that ensures that the text annotation textarea, div, and foreignObject adapt appropriately in terms of height and width when the user types text, drags or resizes the text input.

Reply:
Does these styles affect how we're positing the foreign object? so without the style they would be very different right? Is there a way to include the styles in the created svg? or would that break svg?

Reply:
This affects the sizing of the foreign object. For the first time, these are computed, and they are included in the SVG as height and width, later as attributes/styles of the foreign object. We don't compute them again when the saved annotation gets loaded.